pallets/parachain-staking/src/lib.rs

```diff
@@ -389,287 +458,720 @@ pub mod pallet {
+		/// bottom delegation (and if so, need to adjust the total storage item accordingly fuck)
+		pub fn add_bottom_delegation<T: Config>(
+			&mut self,
+			bumped_from_top: bool,
+			candidate: &T::AccountId,
+			delegation: Bond<T::AccountId, BalanceOf<T>>,
+		) where
+			BalanceOf<T>: Into<Balance>,
+		{
+			let mut bottom_delegations = <BottomDelegations<T>>::get(candidate)
+				.expect("CandidateInfo existence => BottomDelegations existence");
+			// if bottom is full, kick the lowest bottom (which is expected to be lower than input
+			// as per check)
+			let increase_delegation_count = if bottom_delegations.delegations.len() as u32
+				== T::MaxBottomDelegationsPerCandidate::get()
+			{
+				let lowest_bottom_to_be_kicked = bottom_delegations
+					.delegations
+					.pop()
+					.expect("if at full capacity (>0), then >0 bottom delegations exist; qed");
+				assert!(
+					lowest_bottom_to_be_kicked.amount < delegation.amount,
+					"Cannot kick the lowest bottom delegation if the input delegation amount is
+					less than or equal to the current lowest bottom delegation"
+				);
+				bottom_delegations.total = bottom_delegations
+					.total
+					.saturating_sub(lowest_bottom_to_be_kicked.amount);
+				// update delegator state
+				// unreserve kicked bottom
+				T::Currency::unreserve(
+					&lowest_bottom_to_be_kicked.owner,
+					lowest_bottom_to_be_kicked.amount,
+				);
+				// total staked is updated via propagation of lowest bottom delegation amount prior
+				// to call
+				let mut delegator_state =
+					<DelegatorState<T>>::get(&lowest_bottom_to_be_kicked.owner)
+						.expect("Delegation existence => DelegatorState existence");
+				let leaving = delegator_state.delegations.0.len() == 1usize;
+				delegator_state.rm_delegation(candidate);
+				Pallet::<T>::deposit_event(Event::DelegationKicked(
+					lowest_bottom_to_be_kicked.owner.clone(),
+					candidate.clone(),
+					lowest_bottom_to_be_kicked.amount,
+				));
+				if leaving {
+					<DelegatorState<T>>::remove(&lowest_bottom_to_be_kicked.owner);
+					Pallet::<T>::deposit_event(Event::DelegatorLeft(
+						lowest_bottom_to_be_kicked.owner,
+						lowest_bottom_to_be_kicked.amount,
+					));
+				} else {
+					<DelegatorState<T>>::insert(&lowest_bottom_to_be_kicked.owner, delegator_state);
+				}
+				false
+			} else {
+				!bumped_from_top
+			};
+			// only increase delegation count if new bottom delegation (1) doesn't come from top &&
+			// (2) doesn't pop the lowest delegation from the bottom
+			if increase_delegation_count {
+				self.delegation_count += 1u32;
+			}
+			bottom_delegations.insert_sorted_greatest_to_least(delegation);
+			self.reset_bottom_data::<T>(&bottom_delegations);
+			<BottomDelegations<T>>::insert(candidate, bottom_delegations);
+		}
+		/// Remove delegation
+		/// Removes from top if amount is above lowest top or top is not full
+		/// Return Ok(if_total_counted_changed)
+		pub fn rm_delegation<T: Config>(
```

Comment:
Would change this name to better reflect whbat it does, which is a conditional delegation removal

Reply:
Or change the one below to rm_existing_top_delegation

Reply:
renamed to `rm_delegation_if_exists`